src/elf/gnu_hash.rs

```diff
@@ -1,143 +1,206 @@
+            symindex: u32,
+            /// Number of elements in the bloom filter, minus 1
+            maskwords: u32,
+            /// Shift count used in the bloom filter
             shift2: u32,
-            maskbits: u32,
-            bloomwords: &'process [$size], // either 32 or 64 bit masks, depending on platform
-            maskwords_bitmask: u32,
-            buckets: &'process [u32],
-            hashvalues: &'process [u32],
-            symtab: &'process [sym::Sym],
+            /// 2 bit bloom filter on `chains`
+            // Either 32 or 64-bit depending on the class of object
+            bloom_filter: &'a [$size], // => bloom_filter[maskwords]
+            /// GNU hash table bucket array; indexes start at 0. This array holds symbol
+            /// table indexes and contains the index of hashes in `chains`
+            buckets: &'a [u32], // => buckets[nbuckets]
+            /// Hash values; indexes start at 0. This array holds symbol table indexes.
+            chains: &'a [u32], // => chains[total_dynsyms - symindex]
+        }
+
+        impl<'a> core::fmt::Debug for GnuHash<'a> {
+            fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
+                f.debug_struct("GnuHash")
+                    .field("nbuckets", &self.nbuckets)
+                    .field("symindex", &self.symindex)
+                    .field("maskwords", &self.maskwords)
+                    .field("shift2", &self.shift2)
+                    .field("bloom_filter", &self.bloom_filter.as_ptr())
+                    .field("bucket", &self.buckets.as_ptr())
+                    .field("chains", &self.chains.as_ptr())
+                    .finish()
+            }
         }
 
-        impl<'process> GnuHash<'process> {
-            pub unsafe fn new(hashtab: *const u32, total_dynsyms: usize, symtab: &'process [sym::Sym]) -> GnuHash<'process> {
+        impl<'a> GnuHash<'a> {
+            /// Initialize a GnuHash from pointer to `.hash` (or `.gnu.hash` section)
+            /// and total number of dynamic symbols.
+            ///
+            /// # Safety
+            ///
+            /// This function will panic if `maskwords` (from `hashtab`) is not power
+            /// of two.
+            pub unsafe fn new(hashtab: *const u32, total_dynsyms: usize) -> GnuHash<'a> {
                 let nbuckets = *hashtab;
-                let symindex = *hashtab.add(1) as usize;
-                let maskwords = *hashtab.add(2) as usize; // how many words our bloom filter mask has
+                let symindex = *hashtab.add(1);
+                let maskwords = *hashtab.add(2);
                 let shift2 = *hashtab.add(3);
-                let bloomwords_ptr = hashtab.add(4) as *const $size;
-                let buckets_ptr = bloomwords_ptr.add(maskwords) as *const u32;
+                assert!(
+                    (maskwords & (maskwords - 1)) == 0,
+                    "maskwords must be a power of two"
+                );
```

Comment:
Is assert accepted here?

Reply:
I think we should rename this method to `from_raw` or something like that
and return a Result instead of panic in user code.